### PR TITLE
Add resize method to Histogram class

### LINKED ISSUE #17719: Add `resize` method to `Histogram`
Some algorithms (e.g. `WienerSmooth`) change the histogram lengths. A `resize` method can provide the required functionality while ensuring that the sizes `x`, `y`, `z` and `dx` are consistent.

### PR DESCRIPTION
**Pull request by AntonPiccardoSelg**
Allows the user to resize a histogram. This ensures that the the length of x, y, e, and dx are changed consistently.

**To test:**

Code review should be sufficient. Make sure that tests are sensible.

Fixes #17719

_Does not need to be in the release notes._

---
#### Reviewer

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):
##### Code Review
- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)? Is it well structured with small focussed classes/methods/functions?
- [ ] Are there unit/system tests in place? Are the unit tests small and test the a class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?
##### Functional Tests
- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] How do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant documentation been added/updated?
- [ ] Is user-facing documentation written in a user-friendly manner?
- [ ] Has developer documentation been updated if required?
- Does everything look good? Comment with the ship it emoji but don't merge. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/HistogramData/src/Histogram.cpp [hunk 1]
```diff
@@ -253,5 +253,21 @@ template <> void Histogram::checkSize(const BinEdges &binEdges) const {
     throw std::logic_error("Histogram: size mismatch of BinEdges\n");
 }
 
+/** Resets the size of the internal x, dx, y, and e data structures
+
+  The argument refers to the requested y length. All other lengths are
+  then inferred automatically.*/
```
Comment (SimonHeybrock) [inline Framework/HistogramData/src/Histogram.cpp]: Maybe expand the doxygen to mention that the behavior is like `std::vector::resize` (truncating, or appending zeros).

File: Framework/HistogramData/src/Histogram.cpp [hunk 2]
```diff
@@ -253,5 +253,21 @@ template <> void Histogram::checkSize(const BinEdges &binEdges) const {
...
+void Histogram::resize(size_t n) {
+  auto newXSize = xMode() == XMode::Points ? n : n + 1;
+
+  m_x.access().mutableRawData().resize(newXSize);
+  m_y.access().mutableRawData().resize(n);
+  m_e.access().mutableRawData().resize(n);
```
Comment (SimonHeybrock) [inline Framework/HistogramData/src/Histogram.cpp]: There can be histograms that have `m_y == m_e == nullptr`, so probably we should have a check also here (and add a corresponding test).

File: Framework/HistogramData/src/Histogram.cpp [hunk 3]
```diff
@@ -253,5 +253,21 @@ template <> void Histogram::checkSize(const BinEdges &binEdges) const {
...
+
+  if (m_dx.use_count() > 0) {
```
Comment (SimonHeybrock) [inline Framework/HistogramData/src/Histogram.cpp]: We recently added `operator bool()` to `cow_ptr`, so you could simply write `if(m_dx)`, I think.

**Review by SimonHeybrock — COMMENTED** [review]
Looks good apart from comments. Unit tests are good as well.

**Comment by SimonHeybrock** [comment]
Last changes look good, :shipit:!

**Comment by OwenArnold** [comment]
Looks good to me.

**Pull request merged**